src/peft/tuners/adamix.py

```diff
@@ -0,0 +1,416 @@
+            self.expert_mixture[i] = copy.deepcopy(expert)
+        self.num_local_experts = num_local_experts
+
+    def get_expert_by_idx(self, idx):
+        return self.expert_mixture[self.dict_keys[idx]]
+
+    def expert_soup_forward(self, x):
+        output = F.linear(x, self.parameter_dict["weight"], self.parameter_dict["bias"])
+        return output
+
+    def expert_soup(self):
+        self.parameter_dict = {"weight": 0, "bias": 0}
+        for idx in range(self.num_local_experts):
+            single_expert = self.expert_mixture[idx]
+            for s_name, s_param in single_expert.named_parameters():
+                if "weight" in s_name:
+                    p_name = "weight"
+                else:
+                    p_name = "bias"
+                self.parameter_dict[p_name] = self.parameter_dict[p_name] + s_param / self.num_local_experts
+
+    def forward(self, x: torch.Tensor):
+        expert_output = None
+
+        if not self.inference_mode:
+            expert_idx = torch.randint(low=0, high=self.num_local_experts, size=(1,)).item()  # selected expert
+            expert_output = self.get_expert_by_idx(expert_idx)(x)
+
+        else:
+            self.expert_soup()
+            expert_output = self.expert_soup_forward(x)
+
+        return expert_output
+
+
+class ExpertSoup(nn.Module):
+    def __init__(
+        self,
+        hidden_dim,
+        adapter_name,
+        adapter_dim,
+        num_expert=4,
+        sharing_down=False,
+        sharing_up=True,
+        return_two_views=False,
+        inference_mode=False,
+        dtype=torch.float16,
+    ):
+        super().__init__()
+
+        self.disable_adapters = False
+        self.return_two_views = return_two_views
+        self.inference_mode = inference_mode
+        if sharing_down:
+            self.MoA_down = MixtureSoup(nn.Linear(hidden_dim, adapter_dim), adapter_name, self.inference_mode, 1)
+        else:
+            self.MoA_down = MixtureSoup(
+                nn.Linear(hidden_dim, adapter_dim), adapter_name, self.inference_mode, num_expert
+            )
+
+        if sharing_up:
+            self.MoA_up = MixtureSoup(nn.Linear(adapter_dim, hidden_dim), adapter_name, self.inference_mode, 1)
+        else:
+            self.MoA_up = MixtureSoup(
+                nn.Linear(adapter_dim, hidden_dim), adapter_name, self.inference_mode, num_expert
+            )
+
+        self.two_views = []
+        for p in self.parameters():
+            p.data = p.data.to(dtype)
+
+    # NOTE: During training, you must forward pass the input twice to get two outputs and apply the KL div minimization
+    # Use the first ouput as input to next layer
+    # During inference, only one forward pass is required
+    def forward(self, x):
+        if not self.disable_adapters:
+            result1 = F.gelu(self.MoA_down(x))
+            result1 = self.MoA_up(result1)
+            result1 = result1 + x
+
+            if self.training and self.return_two_views:
```

Comment:
Is it possible that `result2` would just use the exact same expert, leading to 0 divergence? If so, is this consistent with the paper? Maybe it doesn't matter in the grand scheme of things.

Reply:
Thats possible, and is consistent with the paper.